File: backend/app/services/region/list.py
"""Region-domain business logic: backend-driven paginated listing.

Mirrors ``services/admin_user/list.py`` and ``services/user/list.py`` exactly —
same pagination pattern, same return signature shape ``tuple[list[DTO], PaginationMeta]``.
All filtering and sorting run in the DB layer; no Python-side post-filtering.
"""

from __future__ import annotations

from math import ceil

from ...core.responses import PaginationMeta
from ...models.region import Region
from ...schemas.region import RegionListQuery, RegionPublic
from ...utils.region.query import build_region_filter, build_sort
from .serialize import to_public


async def list_regions(
    query: RegionListQuery,
) -> tuple[list[RegionPublic], PaginationMeta]:
    """Return a page of regions plus stable pagination metadata.

    Filtering, sorting, and pagination all run in the DB layer so the
    frontend never receives unfiltered data (``backend-api-standards``,
    no client-side filtering).
    """
    filt = build_region_filter(query)

    total = await Region.find(filt).count()

    docs = (
        await Region.find(filt)
        .sort(build_sort(query.sortBy, query.sortOrder))
        .skip(query.skip)
        .limit(query.limit)
        .to_list()
    )

    items = [to_public(doc) for doc in docs]

    meta = PaginationMeta(
        page=query.page,
        limit=query.limit,
        total=total,
        totalPages=ceil(total / query.limit) if query.limit else 0,
        sortBy=query.sortBy,
        sortOrder=query.sortOrder,
    )

    return items, meta

File: frontend/src/api/admin/regions/list.ts
/** `GET /admin/regions` — backend-driven paginated region list. */

import { buildQuery, getList } from "@/api/client"
import type { PaginatedResult } from "@/types/api/envelope"
import type { Region, RegionListQuery } from "@/types/admin/region"

/**
 * Drop "all" sentinel and convert boolean active to string —
 * buildQuery does not accept raw booleans.
 */
function normalize(q: RegionListQuery): Record<string, string | number | undefined> {
  return {
    page: q.page,
    limit: q.limit,
    sortBy: q.sortBy,
    sortOrder: q.sortOrder,
    q: q.q,
    active:
      q.active === "all" ? undefined
      : q.active === true ? "true"
      : q.active === false ? "false"
      : undefined,
  }
}

export function listRegions(query: RegionListQuery = {}): Promise<PaginatedResult<Region>> {
  return getList<Region>(`/admin/regions${buildQuery(normalize(query))}`)
}
